deeppavlov/pipeline_manager/observer.py

```diff
@@ -0,0 +1,180 @@
+                (self.log_path / 'images').mkdir()
+
+        self.exp_info = OrderedDict(date=date,
+                                    exp_name=self.exp_name,
+                                    root=str(self.root),
+                                    info=self.exp_inf,
+                                    number_of_pipes=None,
+                                    metrics=None,
+                                    target_metric=None,
+                                    dataset_name=None)
+        self.log = None
+
+    def save_exp_info(self):
+        """ Write exp_info in json file. """
+        with open(str(self.exp_file), 'w') as exp_file:
+            json.dump(self.exp_info, exp_file)
+
+    def exp_time(self, time: str) -> None:
+        """
+        Adding the time duration of the experiment in log file.
+
+        Args:
+            time: the time duration of the experiment
+
+        Returns:
+            None
+        """
+        with open(str(self.exp_file), 'r') as old_log:
+            old_log = json.load(old_log)
+
+        old_log['full_time'] = time
+        with open(str(self.exp_file), 'w') as log_file:
+            json.dump(old_log, log_file)
+
+    def tmp_reset(self) -> None:
+        """ Reinitialize temporary attributes. """
+        # tmp parameters
+        self.pipe_ind = 0
+        self.pipe_conf = None
+        self.model = None
+        self.pipe_res = None
+        self.pipe_time = None
+        self.batch_size = None
+        self.log = None
+
+    def write(self) -> None:
+        """ Write pipeline logs in jsonl file. """
+        if self.log_file.is_file():
+            with open(str(self.log_file), 'a') as logs:
+                logs.write(json.dumps(self.log))
+                logs.write('\n')
+        else:
+            with open(str(self.log_file), 'w') as logs:
+                logs.write(json.dumps(self.log))
+                logs.write('\n')
+
+    def update_log(self):
+        """ Updates a log with new pipeline information. """
+
+        if (self.model is None) and (self.pipe_conf is not None):
+            for component in self.pipe_conf:
+                if component.get('main') is True:
+                    self.model = component['component_name']
+
+        pipe_name = '-->'.join([x['component_name'] for x in self.pipe_conf])
+
+        self.log = {'pipe_index': self.pipe_ind,
+                    'model': self.model,
+                    'config': self.pipe_conf,
+                    'light_config': pipe_name,
+                    'time': self.pipe_time,
+                    'batch_size': self.batch_size,
+                    'results': self.pipe_res}
+
+        self.write()
+        self.tmp_reset()
+        return self
+
+    def save_config(self, conf: dict, dataset_name: str, ind: int) -> None:
+        """ Save train config in checkpoint folder. """
+        with open(str(self.save_path / dataset_name / "pipe_{}".format(ind) / 'config.json'), 'w') as cf:
```

Comment:
Type casting is not required of opening Paths, you can just `path.open()`
```suggestion
        with self.save_path.joinpath(dataset_name, f'pipe_{ind}', 'config.json').open('w', encoding='utf8') as cf:
```

Reply:
I will know, thank you.